controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,873 @@
+		dnsPrefix = spec.ClusterName
+	}
+
+	tags := make(map[string]*string)
+	for key, val := range spec.Tags {
+		if val != "" {
+			tags[key] = to.StringPtr(val)
+		}
+	}
+	displayName := spec.DisplayName
+	if displayName == "" {
+		displayName = spec.ClusterName
+	}
+	tags["displayName"] = to.StringPtr(displayName)
+
+	kubernetesVersion := spec.KubernetesVersion
+
+	var vmNetSubnetID *string
+	networkProfile := &containersv211.NetworkProfile{}
+	if hasCustomVirtualNetwork(spec) {
+		virtualNetworkResourceGroup := spec.ResourceGroup
+
+		//if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
+		if spec.VirtualNetworkResourceGroup != "" {
+			virtualNetworkResourceGroup = spec.VirtualNetworkResourceGroup
+		}
+
+		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
+			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
+			spec.SubscriptionID,
+			virtualNetworkResourceGroup,
+			spec.VirtualNetwork,
+			spec.Subnet,
+		))
+
+		networkProfile.DNSServiceIP = to.StringPtr(spec.NetworkDNSServiceIP)
+		networkProfile.DockerBridgeCidr = to.StringPtr(spec.NetworkDockerBridgeCIDR)
+		networkProfile.ServiceCidr = to.StringPtr(spec.NetworkServiceCIDR)
+
+		if spec.NetworkPlugin == "" {
+			networkProfile.NetworkPlugin = containersv211.Kubenet
+		} else {
+			networkProfile.NetworkPlugin = containersv211.NetworkPlugin(spec.NetworkPlugin)
+		}
+
+		// if network plugin is 'kubenet', set PodCIDR
+		if networkProfile.NetworkPlugin == containersv211.Azure {
+			networkProfile.PodCidr = to.StringPtr(spec.NetworkPodCIDR)
+		}
+
+		if spec.LoadBalancerSku != "" {
+			loadBalancerSku := containersv211.LoadBalancerSku(spec.LoadBalancerSku)
+			networkProfile.LoadBalancerSku = loadBalancerSku
+		}
+
+		if spec.NetworkPolicy != "" {
+			networkProfile.NetworkPolicy = containersv211.NetworkPolicy(spec.NetworkPolicy)
+		}
+	}
+
+	var agentPoolProfiles []containersv211.ManagedClusterAgentPoolProfile
+	if hasAgentPoolProfile(spec) {
+		for _, np := range spec.NodePools {
+			if np.OrchestratorVersion == "" {
+				np.OrchestratorVersion = spec.KubernetesVersion
+			}
+			agentProfile := containersv211.ManagedClusterAgentPoolProfile{
+				Name:                to.StringPtr(np.Name),
+				Count:               to.Int32Ptr(np.Count),
+				MaxPods:             to.Int32Ptr(np.MaxPods),
+				OsDiskSizeGB:        to.Int32Ptr(np.OsDiskSizeGB),
+				OsType:              containersv211.OSType(np.OsType),
+				VMSize:              containersv211.VMSizeTypes(np.VMSize),
+				Mode:                containersv211.AgentPoolMode(np.Mode),
+				OrchestratorVersion: to.StringPtr(np.OrchestratorVersion),
+			}
+			if len(np.AvailabilityZones) > 0 {
+				agentProfile.AvailabilityZones = to.StringSlicePtr(np.AvailabilityZones)
+			}
+			if hasCustomVirtualNetwork(spec) {
+				agentProfile.VnetSubnetID = vmNetSubnetID
+			}
+			agentPoolProfiles = append(agentPoolProfiles, agentProfile)
+		}
+	} else {
```

Comment:
If spec.NodePools is nil, no changed should take place.

Reply:
Single system node pool is required to specify. This is way I have implemented it in this way.

Reply:
Okay but what about for an imported cluster? Nil fields should indicate that no update is to take place. That is how syncing works. To indicate it is just empty then an empty slice should be used instead.